GR: allow explicit guidefontrotation

t-bltg opened the pull request:
This patch allows using `guidefontrotation` (x or y label rotation) with the GR backend :

**mwe.jl**

```julia
using Plots; gr()

plot(1:2, 1:2, xlabel="Δx", ylabel="ε", yguidefontrotation=-90.)
savefig("foo.png")
```
**output:**
![foo](https://user-images.githubusercontent.com/13423344/120406968-6be48000-c34c-11eb-93b7-54da8dbba0f0.png)

**related:**
https://github.com/JuliaPlots/Plots.jl/issues/3089
https://discourse.julialang.org/t/rotating-the-axes-labels-in-gr/52379

BeastyBlacksmith commented:
can you make it that it rotates around the center of the character? It seems to rotate around the top.

t-bltg commented:
@BeastyBlacksmith here is a patch, but it feels more as a dirty fix than anything else:

```diff
--- gr.jl	2021-06-02 02:08:23.649197290 +0200
+++ gr.jl	2021-06-03 13:00:20.267463879 +0200
@@ -1510,25 +1510,33 @@
         GR.savestate()
         gr_set_font(guidefont(axis), sp)
         guide_position = axis[:guide_position]
+        angle = float(axis[:guidefontrotation])  # github.com/JuliaPlots/Plots.jl/issues/3089
         if letter === :y
-            GR.setcharup(-1, 0)
-            ypos = gr_view_yposition(viewport_plotarea, position(axis[:guidefontvalign]))
-            yalign = alignment(axis[:guidefontvalign])
+            angle += 180.  # default angle = 0. should yield GR.setcharup(-1, 0) i.e. 180°
+            GR.setcharup(cosd(angle), sind(angle))
+            valign = axis[:guidefontvalign]
+            ypos = gr_view_yposition(viewport_plotarea, position(valign))
+            horz_align = alignment(valign)
+            vert = abs(cosd(angle)) > abs(sind(angle))  # is the text more vertical than horizontal ?
             if guide_position === :right || (guide_position == :auto && mirror)
-                GR.settextalign(yalign, GR.TEXT_VALIGN_BOTTOM)
+                GR.settextalign(horz_align, vert ? GR.TEXT_VALIGN_BOTTOM : GR.TEXT_VALIGN_HALF)
                 xpos = viewport_plotarea[2] + 0.03 + mirror * gr_axis_width(sp, axis)
             else
-                GR.settextalign(yalign, GR.TEXT_VALIGN_TOP)
+                GR.settextalign(horz_align, vert ? GR.TEXT_VALIGN_TOP : GR.TEXT_VALIGN_HALF)
                 xpos = viewport_plotarea[1] - 0.03 - !mirror * gr_axis_width(sp, axis)
             end
         else
-            xpos = gr_view_xposition(viewport_plotarea, position(axis[:guidefonthalign]))
-            xalign = alignment(axis[:guidefonthalign])
+            angle += 90.  # default angle = 0. should yield GR.setcharup(0, 1) i.e. 90°
+            GR.setcharup(cosd(angle), sind(angle))
+            halign = axis[:guidefonthalign]
+            xpos = gr_view_xposition(viewport_plotarea, position(halign))
+            horz_align = alignment(halign)
+            vert = abs(cosd(angle)) > abs(sind(angle))
             if guide_position === :top || (guide_position == :auto && mirror)
-                GR.settextalign(xalign, GR.TEXT_VALIGN_TOP)
+                GR.settextalign(horz_align, vert ? GR.TEXT_VALIGN_HALF : GR.TEXT_VALIGN_TOP)
                 ypos = viewport_plotarea[4] + 0.015 + (mirror ? gr_axis_height(sp, axis) : 0.015)
             else
-                GR.settextalign(xalign, GR.TEXT_VALIGN_BOTTOM)
+                GR.settextalign(horz_align, vert ? GR.TEXT_VALIGN_HALF : GR.TEXT_VALIGN_BOTTOM)
                 ypos = viewport_plotarea[3] - 0.015 - (mirror ? 0.015 : gr_axis_height(sp, axis))
             end
         end
```

Also, some hard-coded constants don't take into account the possibility of the text being horizontal on the `y` axis:
https://github.com/JuliaPlots/Plots.jl/blob/master/src/backends/gr.jl#L1519.

Maybe this will need a full rewrite, that's why I'm not in favor of breaking the current bahavior.

BeastyBlacksmith commented:
When it comes to the GR backend, I am all in favor of rewriting stuff such that we don't have so many magic constants or using higher level interfaces.
However, I can understand if you don't want to put so much work in it and I am going to merge this as is to get at least some rotation.
You can then tackle a rewrite in a different PR if you want to.

Before that, I'd like to ask you to add your information to our `.zenodo.json` file, such that you get proper authorship credit. Thank you for contributing to Plots.jl :)

t-bltg commented:
Agreed, let's proceed incrementally and rewrite this later, when I will have more experience with the `Plots.jl` ecosystem and the `GR` backend.
I've added the credit and squashed the 2 commits.
Thanks !
